Assets/__Scripts/UI/SongSelectMenu/SongListItem.cs

```diff
@@ -1,27 +1,341 @@
-﻿using TMPro;
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using SimpleJSON;
+using TMPro;
 using UnityEngine;
+using UnityEngine.EventSystems;
+using UnityEngine.Networking;
 using UnityEngine.UI;
 
-public class SongListItem : MonoBehaviour {
+[RequireComponent(typeof(Image))]
+public class SongListItem : RecyclingListViewItem, IPointerEnterHandler, IPointerExitHandler, IPointerClickHandler {
 
-    [SerializeField]
-    TextMeshProUGUI text;
+    private SongList _songList;
+    
+    [SerializeField] private TextMeshProUGUI title;
+    [SerializeField] private TextMeshProUGUI artist;
+    [SerializeField] private TextMeshProUGUI folder;
 
-    [SerializeField]
-    Button button;
+    [SerializeField] private TextMeshProUGUI duration;
+    [SerializeField] private TextMeshProUGUI bpm;
 
-    private BeatSaberSong song;
+    [SerializeField] private Image cover;
+    [SerializeField] private Sprite defaultCover;
+    
+    [SerializeField] private GameObject rightPanel;
+    [SerializeField] private Toggle favouriteToggle;
+    private Image _bg;
+    private string _previousSearch = "";
 
-    public void AssignSong(BeatSaberSong song) {
-        this.song = song;
-        button.onClick.RemoveAllListeners();
-        button.onClick.AddListener(ButtonClicked);
-        text.text = $"{song.songName.StripTMPTags()} <size=50%><i>{song.songSubName.StripTMPTags()}</i></size>";
+    private BeatSaberSong _song;
+    private static readonly Dictionary<string, WeakReference<Sprite>> Cache = new Dictionary<string, WeakReference<Sprite>>();
+    private static readonly Dictionary<string, float> DurationCache = new Dictionary<string, float>();
+    private static bool _hasAppliedThisFrame = false;
+    
+    private static string _durationCachePath;
+    private static JSONObject _songCoreCache;
+
+    private bool _ignoreToggle = false;
+
+    private void Start()
+    {
+        rightPanel.SetActive(false);
+        _bg = GetComponent<Image>();
+        // I have sinned
+        _songList = FindObjectOfType<SongList>();
```

Comment:
excuse me what the fuck is this

![image](https://user-images.githubusercontent.com/30004286/118339816-d41f0f00-b4ce-11eb-826c-96fba7a45e3c.png)